app/uk/gov/hmrc/breathingspaceifstub/controller/AbstractBaseController.scala

```diff
@@ -98,4 +111,46 @@ abstract class AbstractBaseController(cc: ControllerComponents) extends BackendC
           }
         )
     }
+
+  private def validateCorrelation(headers: Headers): Either[HttpError, String] =
+    headers
+      .get(Header.CorrelationId)
+      .fold[Either[HttpError, String]](
+        Left(HttpError(none, Failure(MISSING_HEADER)))
+      )(Right(_))
+
+  private def validateOriginatorId(headers: Headers, endpointId: EndpointId)(
+    implicit correlationId: String
+  ): Either[HttpError, Boolean] =
+    headers
+      .get(Header.OriginatorId)
+      .fold[Either[HttpError, Boolean]](
+        Left(HttpError(correlationId.some, Failure(MISSING_HEADER)))
+      ) { originatorId =>
+        Attended.withNameOption(originatorId.toUpperCase) match {
+          case Some(Attended.DA2_BS_ATTENDED) =>
+            if (endpointId == BS_Periods_POST || endpointId == BS_Periods_PUT) {
+              Left(
+                HttpError(correlationId.some, Failure(INVALID_HEADER))
+              )
+            } else Right(true) // UserId is required
+
+          case Some(Attended.DA2_BS_UNATTENDED) => Right(false) // UserId is not required
+
+          case _ => Left(HttpError(correlationId.some, Failure(INVALID_HEADER)))
+        }
+      }
+
+  private def validateUserId(headers: Headers, isUserIdRequired: Boolean)(
+    implicit correlationId: String
+  ): Either[HttpError, Unit] =
+    headers
+      .get(Header.UserId)
+      .fold[Either[HttpError, Unit]] {
+        if (isUserIdRequired) Left(HttpError(correlationId.some, Failure(MISSING_HEADER)))
+        else Right(unit)
+      } { _ =>
+        if (isUserIdRequired) Right(unit)
+        else Left(HttpError(correlationId.some, Failure(INVALID_HEADER)))
```

Comment:
Nit: Rather than a fold with two if else blocks you could replace with a match statement for improved readability (IMO) :slightly_smiling_face: